common/src/main/java/org/apache/celeborn/common/network/sasl/registration/RegistrationClientBootstrap.java

```diff
@@ -0,0 +1,260 @@
+   * TODO: This should be made configurable. For now, we only support ANONYMOUS for client-auth and
+   * DIGEST-MD5 for connect-auth.
+   */
+  private static final List<PbSaslMechanism> SASL_MECHANISMS =
+      Lists.newArrayList(
+          PbSaslMechanism.newBuilder()
+              .setMechanism(ANONYMOUS)
+              .addAuthTypes(PbAuthType.CLIENT_AUTH)
+              .build(),
+          PbSaslMechanism.newBuilder()
+              .setMechanism(DIGEST_MD5)
+              .addAuthTypes(PbAuthType.CONNECTION_AUTH)
+              .build());
+
+  private final TransportConf conf;
+  private final String appId;
+  private final SaslCredentials saslCredentials;
+
+  private final RegistrationInfo registrationInfo;
+
+  public RegistrationClientBootstrap(
+      TransportConf conf,
+      String appId,
+      SaslCredentials saslCredentials,
+      RegistrationInfo registrationInfo) {
+    this.conf = Preconditions.checkNotNull(conf, "conf");
+    this.appId = Preconditions.checkNotNull(appId, "appId");
+    this.saslCredentials = Preconditions.checkNotNull(saslCredentials, "saslCredentials");
+    this.registrationInfo = Preconditions.checkNotNull(registrationInfo, "registrationInfo");
+  }
+
+  @Override
+  public void doBootstrap(TransportClient client, Channel channel) throws RuntimeException {
+    if (registrationInfo.getRegistrationState() == RegistrationInfo.RegistrationState.REGISTERED) {
+      LOG.info("client has already registered, skip register.");
+      doSaslBootstrap(client, channel);
+      return;
+    }
+    try {
+      LOG.info("authentication initiation started for {}", appId);
+      doAuthInitiation(client);
+      LOG.info("authentication initiation successful for {}", appId);
+      doClientAuthentication(client);
+      LOG.info("client authenticated for {}", appId);
+      register(client);
+      LOG.info("Registration for {}", appId);
+      registrationInfo.setRegistrationState(RegistrationInfo.RegistrationState.REGISTERED);
+    } catch (Throwable e) {
+      LOG.error("Registration failed for {}", appId, e);
+      registrationInfo.setRegistrationState(RegistrationInfo.RegistrationState.FAILED);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void doAuthInitiation(TransportClient client) throws IOException, CelebornException {
+    PbAuthenticationInitiationRequest authInitRequest =
+        PbAuthenticationInitiationRequest.newBuilder()
+            .setVersion(VERSION)
+            .setAuthEnabled(true)
+            .addAllSaslMechanisms(SASL_MECHANISMS)
+            .build();
+    TransportMessage msg =
+        new TransportMessage(
+            MessageType.AUTHENTICATION_INITIATION_REQUEST, authInitRequest.toByteArray());
+    ByteBuffer authInitResponseBuffer;
+    try {
+      authInitResponseBuffer = client.sendRpcSync(msg.toByteBuffer(), conf.saslTimeoutMs());
+    } catch (RuntimeException ex) {
+      // TODO: Auth initiation timed out. Will just throw SaslTimeoutException for now
```

Comment:
Throwing `SaslTimeoutException` is sufficient, right ? What is the change the TODO is for ?